### PR TITLE
Check reachability when looking at uninitialized non-null fields

### LINKED ISSUE #25529: Reporting "field is uninitialized" for constructor that throws an exception
Reports: `(5,14): warning CS8618: Non-nullable field '_f' is uninitialized.`
```
using System;
class C
{
    private object _f;
    internal C()
    {
        throw new NotImplementedException();
    }
}
```

### PR DESCRIPTION
**Pull request by agocke**
If the end of a constructor is unreachable we shouldn't warn about
fields which are uninitialized at the end.

Fixes #25529

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -85,6 +85,41 @@ public override string ToString()
     }
 }";
 
+        [Fact]
+        [WorkItem(25529, "https://github.com/dotnet/roslyn/issues/25529")]
+        public void UnassignedNonNullFieldsUnreachable()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >UnassignedNonNullFieldsUnreachable [](start = 20, length = 34)

Consider moving to `UninitializedNonNullableFieldTests`.

**Comment by gafter** [comment]
>     internal sealed class UnassignedFieldsWalker : DefiniteAssignmentPass

This code should be integrated into the definite assignment pass.  There is no need for the compiler to repeat all that work.

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/UnassignedFieldsWalker.cs:17 in 971fcbb. [](commit_id = 971fcbb03a49664a2e2b4c4d18cba5cbcf82ce3e, deletion_comment = False)

**Comment by gafter** [comment]
>             ReportUninitializedNonNullableReferenceTypeFields();

This only performs the analysis on the "fall through" path.  You also need to perform the analysis on each return statement (a list of which are available in "result")

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/UnassignedFieldsWalker.cs:63 in 971fcbb. [](commit_id = 971fcbb03a49664a2e2b4c4d18cba5cbcf82ce3e, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Also please add tests that have return statements inside the constructor (at the end, and inside of an if statement)

---
In reply to: [267909107](https://github.com/dotnet/roslyn/pull/34316#discussion_r267909107) [](ancestors = 267909107)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -85,6 +85,41 @@ public override string ToString()
...
+                // (5,20): warning CS0169: The field 'C._f' is never used
+                //     private object _f;
+                Diagnostic(ErrorCode.WRN_UnreferencedField, "_f").WithArguments("C._f").WithLocation(5, 20),
+                // (11,14): warning CS8618: Non-nullable field '_f' is uninitialized.
+                //     internal C(object o) { }
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "_f").WithLocation(11, 14),
+                // (13,14): warning CS8618: Non-nullable field '_f' is uninitialized.
+                //     internal C(string s)
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "_f").WithLocation(13, 14));
+        }
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: This diagnostic appears to be incorrect.  At the end point of this ctor the field _f is definitely assigned.

Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: This diagnostic should preferably point to the return statement.  I believe that's what we do for structs.

**Comment by jcouv** [comment]
It looks like https://github.com/dotnet/roslyn/issues/30022 may have been closed prematurely (it's closed already).

**Comment by agocke** [comment]
@jcouv I don't think so, that one's specifically about struct fields which still shouldn't be checked because definite assignment is stricter. The only  remaining question is class fields, which I'm working on fixing now by merging this with definite assignment.

**Comment by jcouv** [comment]
> It looks like #30022 may have been closed prematurely (it's closed already).

I was observing that this PR says "Fixes #30022", but that issue is already closed.

**Comment by agocke** [comment]
@jcouv Thanks! I was linking the wrong bug.

@gafter I've updated the tests. I think this PR implements the correct behavior, just with worse diagnostics and performance. The "fall-through" state in this case is the Joined state for all of the return branches, so I think it will never fail to report an unintialized field. I've filed https://github.com/dotnet/roslyn/issues/34598 to track the larger work of integrating with the DefiniteAssignment pass directly.

**Pull request merged**